pkg/kubelet/dockershim/exec.go

```diff
@@ -109,53 +105,55 @@ func (*NativeExecHandler) ExecInContainer(client libdocker.Interface, container
 		RawTerminal:  tty,
 		ExecStarted:  execStarted,
 	}
-	err = client.StartExec(execObj.ID, startOpts, streamOpts)
-	if err != nil {
-		return err
+
+	if timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
 	}
 
-	// if ExecProbeTimeout feature gate is disabled, preserve existing behavior to ignore exec timeouts
-	var execTimeout <-chan time.Time
-	if timeout > 0 && utilfeature.DefaultFeatureGate.Enabled(features.ExecProbeTimeout) {
-		execTimeout = time.After(timeout)
-	} else {
-		// skip exec timeout if provided timeout is 0
-		execTimeout = nil
+	// StartExec is a blocking call, so we need to run it concurrently and catch
+	// its error in a channel
+	execErr := make(chan error, 1)
+	go func() {
+		execErr <- client.StartExec(execObj.ID, startOpts, streamOpts)
+	}()
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-execErr:
+		if err != nil {
```

Comment:
nothing bad will happen, but maybe need to call `cancel()`?

Reply:
cancel is called with defer on line 112

Reply:
yes, totally. My comment is that we can cancel timeout in case the exec has already finished =)